### ISSUE TITLE
Meta data updates does not update during kobo sync

### ISSUE THREAD

**Issue by joggs** [janeczku/calibre-web#2816]
**Describe the bug/problem**
Meta data updates does not update during kobo sync. you always find a missing description etc and changes it in calibre, and then want to sync it the calibre-web-> kobo way. but it does not get updated.

**To Reproduce**
1. have a fully synced kobo/calibre-web setup
2. edit the metadata for a book in calibre
3. sync with kobo
4. the changes are not updated

**Logfile**
Add content of calibre-web.log file or the relevant error, try to reproduce your problem with "debug" log-level to get more output.

**Expected behavior**
the updated metadata should be synced to the kobo

**Environment (please complete the following information):**
 - OS: docker linux(unraid)
 - Python version: 3.10.6 (main, May 29 2023, 11:10:38) [GCC 11.3.0]
 - Calibre-Web version: 0.6.21 Beta
 - Docker container: [LinuxServer]:
 - Special Hardware: 

**Additional context**
same results using reverse proxy(traefik) as when using direct access.

**Comment by OzzieIsaacs** [comment]
Your obersation is comprehensible. This is a limitation of the sync protocol and the kobo reader. As long as the book Id (more precise the uuid of the book in the calibre database) isn't changed, a new sync doesn't change anythin on the kobo side.
I think there is already another issue about this

**Comment by joggs** [comment]
Thank you for your quick answer. I could not find a previous issue regarding this, but I'll have another look, as there could be ideas of workarounds. Maybe I'll try to modify the uuid directly in the database to see what happens... or if you could get an editable column with uuid in calibre? :-) I have to dive deeper into calibre and calibre-web

**Comment by fancl20** [comment]
I think the root cause is same as #2509. My understanding is we don't have to use the uuid of the book in the calibre database as the id in the kobo protocol. A computed id (from the content of metadata) can trick kobo think it's a different book.

Though I don't know how difficult the change can be :/

**Comment by smellslikemoose** [comment]
Still an issue in Feb 2024, wondering if there’s been any progress?

**Comment by FerdiZwanink** [comment]
Hiya, any progress on this issue? My work around so far was t make sure all metadata is complete before sync. Created a clear 'zero point' by continually forcing the whole database to sync, which is tedious to say the least. :D

**Comment by tsheinen** [comment]
I'd like to prefix this with "I've done no testing at all just some vibes-based libnickel reverse engineering". I haven't fully followed through how Kobo processes these but I did see some potentially telling strings in the part of the library that handles the server response. 

![image](https://github.com/user-attachments/assets/e9307160-746c-4e4f-9490-4f72173cca45)

![image](https://github.com/user-attachments/assets/084835f1-e168-4342-9966-0db5b65b6f33)

It looks like the sync protocol does have an instruction, "ChangedProductMetadata", to invalidate/force resync book metadata? The JSON format it expects is unclear to me right now but I think this is possible -- it's just a matter of understanding the protocol better.  I'll keep poking around libnickel and see if I can figure out in more detail how to do this.

**Comment by jmarmstrong1207** [comment]
> Okay, actually it was really straightforward!
> 
> I did a brief patch to emit ChangedProductMetadata ([tsheinen@a4781a6](https://github.com/tsheinen/calibre-web/commit/a4781a62b2fde549a40b87ccabb6b0df20ce60a4)) for every synced book and did observe metadata set in the calibre-web UI propagating to Kobo. There seems to be minor differences in the metadata format between a new book and changing metadata but I don't think it's actually a problem because it just ignores extra fields.
> 
> syslog had one of these for each synced book
> 
> ```
> Aug 25 18:44:04 nickel: (  4440.204 @ 0x2078ab0 / ui.warning) void LibraryParser::parseChangedMetadata(const QVariantMap&) Ignoring unrecognized metadata "Categories"
> ```

Would this also cover book cover changes as well?

**Comment by jmarmstrong1207** [comment]
@tsheinen If it also works, you definitely should pull request the patch to this repository!

**Comment by tsheinen** [comment]
in my testing it works reliably -- I just haven't had the time to put together a good patch. I did a quick and dirty patch to test how kobo responds to that sync command but it's super lazy -- just emitting a changed metadata command for everything in the library. A proper change would compare the metadata change timestamp (which i believe is present in the calibre database?) to the last kobo sync time (which i believe is present in the sync token) and only emit the command as necessary. 

I haven't done any testing or investigation on if this command also includes covers. I would imagine it would? Although I'm not sure if it'll just do it automatically or if it'll require some playing with `CoverImageId`. It seems like in the existing implementation it just reuses the book uuid but it's unclear to me if this correlation is important or just a way to implicitly guarantee that each cover uuid is in fact unique. I can imagine a few plausible ways libnickel would handle this (or who knows, it could be different, this is just some educated guessing)

- ChangeProductMetadata does cache invalidation and changing the cover will Just Work 
- cache fetch uses `CoverImageId` and it just needs to be unique (idk hash the cover?)
- cache fetch uses Some Other Information (in particular i'm curious if it keys on `RevisionId` for anything)

tl;dr idk i'll get a patch done eventually but i need to be able to clear an evening to write a good patch and test it and that's depressingly difficult

**Comment by dphildebrandt** [comment]
I'd rather have a quick'n'dirty patch that works (inefficiently) right now, than a perfect one at some possible future time. 

Think of it as a minimum viable product, or, don't let perfect be the enemy of good.😆

**Comment by tsheinen** [comment]
Relevant note: `ChangedProductMetadata` must be combined with `NewEntitlement` to propagate. It's unclear to me what the difference between `NewEntitlement` and `ChangedEntitlement` since everything I've tested seems to work fine with `NewEntitlement`

**Comment by FerdiZwanink** [comment]
@tsheinen any updates on this one?

**Comment by AsherMaximum** [comment]
I can confirm that `ChangedEntitlement` does work for removing books, by setting `IsRemoved `in the entitlement details to True.
It can also be used to tell the Kobo to get a new version of the book file, by changing `RevisionId`. Interestingly, it seems to keep to old file when it does this, although it is inaccessable to the Kobo. 
Changing the `RevisionId` does not update the metadata though.

I think `ChangedEntitlement`  is meant for changing anything that is in the `BookEntitlement` section.

> Relevant note: `ChangedProductMetadata` must be combined with `NewEntitlement` to propagate. It's unclear to me what the difference between `NewEntitlement` and `ChangedEntitlement` since everything I've tested seems to work fine with `NewEntitlement`

This means that `ChangedProductMetadata` isn't being used, it's just getting the new metadata from `NewEntitlement`.

The reason you found it not to work was cause it's not quite right. You need to send `ChangedProductMetadata` along with the entitlement details and the metadata just like `NewEntitlement`:

        log.debug("Syncing book %s, ts_created: %s", book.Books.id, ts_created)
        if ts_created > sync_token.books_last_created:
            entitlement["BookMetadata"] = get_metadata(book.Books)
            sync_results.append({"NewEntitlement": entitlement})
        else:
            entitlement["BookMetadata"] = get_metadata(book.Books)
            sync_results.append({"ChangedProductMetadata": entitlement})

This successfully causes metadata to get updated for me (Kobo Libra 2, kobo version: 4.38.23171).

That is, when it's sent - I'm working on the final touches to it, but I was working on updates to send a `ChangedEntitlement` whenever a book has been updated. It wasn't working, so I found this thread and ran a bunch of tests with various ways of sending `ChangedProductMetadata` until I got it working.

I'll post a PR here once I've finalized the changes and merged them out of my dev branch into their own.

**Comment by AsherMaximum** [comment]
I added my changes to an existing PR I have for handling books removed from a Kobo sync shelf via the app.
Now the PR covers a Kobo sync for both removed and changed books.

**Comment by new-usemame** [comment]
If this is still hitting anyone here, [Calibre-Web-NextGen](https://github.com/new-usemame/Calibre-Web-NextGen) (a community-maintained CWA build) ships a fix for it:

```
docker pull ghcr.io/new-usemame/calibre-web-nextgen:latest
```

What changed:

- `Books.atom_timestamp` now returns `last_modified` (the column that updates on every metadata edit) instead of `timestamp` (date added, set once at import and never touched again). Kobo's sync decides what's new from `atom:updated`; when that field never moved, post-import edits stayed invisible to the device.
- Shipped in v4.0.18 ([`2379867c`](https://github.com/new-usemame/Calibre-Web-NextGen/commit/2379867c)), backport of @haraldpdl's patch ([`ab17992e`](https://github.com/janeczku/calibre-web/commit/ab17992e)). Multi-device sync edge cases are tightened further in v4.0.67 ([`ae3e1936`](https://github.com/new-usemame/Calibre-Web-NextGen/commit/ae3e1936)).

Edit the metadata in Calibre (or directly in the web UI), trigger a sync from the Kobo, and the device should pick up the new cover / description / tags on the next pull.

If you switch over and still see edits not propagating, the fork tracker is at https://github.com/new-usemame/Calibre-Web-NextGen/issues. Heads-up: this build is CWA-derived, so it also bakes in the auto-ingest folder watcher, kepubify pipeline, and KOReader sync on top of the calibre-web core — fine if you want those, worth knowing if you don't.

**Comment by mirthquake** [comment]
Just ran into this issue of updated metadata in Calibre-Web not being synced across to my Kobo Clara BW. I noticed this discussion's been going on for a few years now, so wanted to ask if this feature has actually been implemented yet. FWIW, I'm on the nightly build of the Calibre-Web docker container. Happy to provide any other information if needed. Cheers!
